libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+				break;
+			}
+		}
+	}
+	return ret;
+}
+
+RList *w32_thread_list(RDebug *dbg, int pid, RList *list) {
+	// pid is not respected for TH32CS_SNAPTHREAD flag
+	HANDLE th = CreateToolhelp32Snapshot (TH32CS_SNAPTHREAD, 0);
+	if(th == INVALID_HANDLE_VALUE) {
+		r_sys_perror ("w32_thread_list/CreateToolhelp32Snapshot");
+		return list;
+	}
+	THREADENTRY32 te;
+	te.dwSize = sizeof (te);
+	if (Thread32First (th, &te)) {
+		// TODO: export this code to its own function?
+		HANDLE ph = OpenProcess (PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, pid);
+		char *path = NULL;
+		int uid = -1;
+		if (ph != (HANDLE)NULL) {
+			path = resolve_path (ph);
+			DWORD sid;
+			if (ProcessIdToSessionId (pid, &sid)) {
+				uid = sid;
+			}
+			CloseHandle (ph);
+		}
+		if (!path) {
+			// TODO: enum processes to get binary's name
+			path = strdup ("???");
+		}
+		do {
+			if (te.th32OwnerProcessID == pid) {
+				// TODO: add pc if process is debugged
+				/*ut64 pc;
+				if (dbg->pid == pid) {
+					CONTEXT ctx = {0};
+					w32_reg_read (dbg, R_REG_TYPE_GPR, (ut8 *)&ctx, sizeof (ctx));
+					pc = ctx->eip;
+				}*/
+				r_list_append (list, r_debug_pid_new (path, te.th32ThreadID, uid, 's', 0));
+			}
+		} while (Thread32Next (th, &te));
+		free (path);
+	} else {
+		r_sys_perror ("w32_thread_list/Thread32First");
+	}
+	CloseHandle (th);
+	return list;
+}
+
+static void w32_info_user(RDebug *dbg, RDebugInfo *rdi) {
+	HANDLE h_tok = NULL;
+	DWORD tok_len = 0;
+	PTOKEN_USER tok_usr = NULL;
+	LPTSTR usr = NULL, usr_dom = NULL;
+	DWORD usr_len = 512;
+	DWORD usr_dom_len = 512;
+	SID_NAME_USE snu = {0};
+	RIOW32Dbg *rio = dbg->user;
+
+	if (!OpenProcessToken (rio->ph, TOKEN_QUERY, &h_tok)) {
+		r_sys_perror ("w32_info_user/OpenProcessToken");
+		goto err_w32_info_user;
+	}
+	if (!GetTokenInformation (h_tok, TokenUser, (LPVOID)&tok_usr, 0, &tok_len) && GetLastError () != ERROR_INSUFFICIENT_BUFFER) {
+		r_sys_perror ("w32_info_user/GetTokenInformation");
+		goto err_w32_info_user;
+	}
+	tok_usr = (PTOKEN_USER)malloc (tok_len);
+	if (!tok_usr) {
+		perror ("w32_info_user/malloc tok_usr");
+		goto err_w32_info_user;
+	}
+	if (!GetTokenInformation (h_tok, TokenUser, (LPVOID)tok_usr, tok_len, &tok_len)) {
+		r_sys_perror ("w32_info_user/GetTokenInformation");
+		goto err_w32_info_user;
+	}
+	usr = (LPTSTR)malloc (usr_len);
+	if (!usr) {
+		perror ("w32_info_user/malloc usr");
+		goto err_w32_info_user;
+	}
+	*usr = '\0';
+	usr_dom = (LPTSTR)malloc (usr_dom_len);
+	if (!usr_dom) {
+		perror ("w32_info_user/malloc usr_dom");
+		goto err_w32_info_user;
+	}
+	*usr_dom = '\0';
+	if (!LookupAccountSid (NULL, tok_usr->User.Sid, usr, &usr_len, usr_dom, &usr_dom_len, &snu)) {
+		r_sys_perror ("w32_info_user/LookupAccountSid");
+		goto err_w32_info_user;
+	}
+	if (*usr_dom) {
+		rdi->usr = r_str_newf (W32_TCHAR_FSTR"\\"W32_TCHAR_FSTR, usr_dom, usr);		
+	} else {
+		rdi->usr = r_sys_conv_win_to_utf8 (usr);
+	}
+err_w32_info_user:
+	if (h_tok) {
```

Comment:
is this really needed? `CloseHandle(NULL)` is not like `free(NULL)` ?

Reply:
Calling CloseHandle with invalid handle is kind of ugly thing to do and also throws exception if the process doing so is debugged. (its not even documented)